Homework2/CoherentTrainingSystem/Training.cs

```diff
@@ -0,0 +1,36 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+
+namespace CoherentTrainingSystem
+{
+    public class Training : TrainingBase
+    {
+        public List<TrainingBase> LecturesAndPractice = new List<TrainingBase>();
+
+        public void Add(TrainingBase training)
+        {
+            LecturesAndPractice.Add(training);
+        }
+
+        public void Add(List<TrainingBase> trainings)
+        {
+            LecturesAndPractice.AddRange(trainings);
+        }
+
+        public bool IsPractical()
+        {
+            return LecturesAndPractice.All(l => l is PracticalLesson);
+        }
+
+        public Training Clone()
+        {
+            // Creating a new object for a deep cloning
+            Training deepcopyTraining = new Training();
+
+            // Copying not the reference but the values itself
+            deepcopyTraining.LecturesAndPractice.AddRange(this.LecturesAndPractice);
```

Comment:
Deep clone assumes that each element of LecturesAndPractice is cloned. 
How it should be:

```
Training Clone()
{
     Training clone = new Training(this.Description);
     foreach(var item in this.TrainingActivities)
     {
         clone.Add(item.Clone());
     }
     return clone;
}
// I do not provide a code for cloning PracticalLesson / Lecture
```